### PR TITLE
Add date picker range limit feedback

### PR DESCRIPTION
**Pull request by pochretien**
Add the possibility to enter to type of format for the time in the date input
Add a range limit to show a custom message in a tooltip and disabled the button if exceeded
Fix all ts error in the UTs
Add UTs to convert at 100% the new change
Fix tu current UTs after all changes

### CODE REVIEW

File: src/components/button/Button.tsx [hunk 1]
```diff
@@ -51,23 +56,21 @@ export class Button extends React.Component<IBaseActionOptions, {}> {
     }
 
     return !_.isEmpty(this.props.tooltip)
-      ? <Tooltip title={this.props.tooltip} placement={this.props.tooltipPlacement}>
-        <span>
-          {buttonElement}
-        </span>
+      ? <Tooltip title={this.props.tooltip} placement={this.props.tooltipPlacement} className={'btn-container'}>
```
Comment (sroy3) [inline src/components/button/Button.tsx]: `className` can just be a string

File: src/components/calendar/CalendarConnected.tsx [hunk 2]
```diff
@@ -10,9 +10,8 @@ import {
 import {
   changeDatePickerLowerLimit,
   changeDatePickerUpperLimit,
-  DateLimits,
-  resetDatePickers,
-  selectDate
+  DateLimits, resetDatePickers,
+  selectDate,
```
Comment (sroy3) [inline src/components/calendar/CalendarConnected.tsx]: Optimize imports

File: src/components/calendar/tests/CalendarConnected.spec.tsx [hunk 3]
```diff
@@ -197,7 +197,7 @@ describe('Calendar', () => {
       let currentLowerLimit: Date = moment(new Date()).add(10, 'day').toDate();
       let newLimit: Date = moment(new Date()).subtract(5, 'day').toDate();
 
-      store.dispatch(addDatePicker(PICKER_ID, true, 'any', CALENDAR_ID));
+      store.dispatch(addDatePicker(PICKER_ID, true, undefined, 'any', CALENDAR_ID));
       store.dispatch(changeDatePickerLowerLimit(PICKER_ID, currentLowerLimit));
```
Comment (sroy3) [inline src/components/calendar/tests/CalendarConnected.spec.tsx]: Maybe we should make this a function so next time an argument changes, we won't have to change it 20 times.

File: src/components/datePicker/DatePickerBox.tsx [hunk 4]
```diff
@@ -14,6 +14,7 @@ export interface IDatesSelectionBox {
   title: string;
   quickOptions?: IOption[];
   isRange?: boolean;
+  rangeLimit?: IRangeLimit;
```
Comment (sroy3) [inline src/components/datePicker/DatePickerBox.tsx]: I'm wondering if you could do something with the `selectionRules` prop a few lines down instead. Sometimes, validation is more complex than a simple range limit and it kind of duplicates the rules.

File: src/components/datePicker/DatePickerDropdown.tsx [hunk 5]
```diff
@@ -135,7 +138,19 @@ export class DatePickerDropdown extends React.Component<IDatePickerDropdownProps
       : DateUtils.getSimpleDate(date);
   }
 
+  private hasExceededRangeLimit(): Boolean {
```
Comment (sroy3) [inline src/components/datePicker/DatePickerDropdown.tsx]: Once again, you could validate the `selectionRules` prop instead.

File: src/components/datePicker/DatesSelectionConnected.tsx [hunk 6]
```diff
@@ -29,6 +29,8 @@ const mapStateToProps = (state: IReactVaporState, ownProps: IDatesSelectionOwnPr
   return {
     lowerLimit: item ? item.lowerLimit : new Date(),
     upperLimit: item ? item.upperLimit : new Date(),
+    inputLowerLimit: item ? item.inputLowerLimit : undefined,
+    inputUpperLimit: item ? item.inputUpperLimit : undefined,
```
Comment (sroy3) [inline src/components/datePicker/DatesSelectionConnected.tsx]: You could change this to simply `item && item.inputLowerLimit` and `item && item.inputUpperLimit`

**Comment by sroy3** [comment]
Good job overall! I think this will make the date picker a lot more enjoyable

File: src/components/datePicker/DatePicker.tsx [hunk 7]
```diff
@@ -52,22 +51,21 @@ export class DatePicker extends React.Component<IDatePickerProps, any> {
...
+    this.handleChangeDate();
   }
 
-  private handleChange() {
+  private handleChangeDate() {
     const date: Date = this.getDateFromString(this.dateInput.value);
 
     if (date.getDate()) {
-      this.props.onChange(date, this.props.upperLimit);
+      this.props.onBlur(date, this.props.upperLimit);
```
Comment (MichelML) [inline src/components/datePicker/DatePicker.tsx]: suppose I change the value in the input, and than click on "Apply" without clicking elsewhere, my click won't work, only the second one and up will work. Could a debounced function work here?

File: src/components/datePicker/DatePicker.tsx [hunk 8]
```diff
@@ -107,7 +105,7 @@ export class DatePicker extends React.Component<IDatePickerProps, any> {
         <input
           className={inputClasses.join(' ')}
           ref={(dateInput: HTMLInputElement) => this.dateInput = dateInput}
-          onChange={() => this.handleChange()}
+          onBlur={() => this.handleChangeDate()}
```
Comment (MichelML) [inline src/components/datePicker/DatePicker.tsx]: would be curious to see if `onChange={_.debounce(() => this.handleChangeDate(), 500)}` work here

**Pull request merged**